primitives/src/merkle_tree/field_based_mht/optimized/mod.rs

```diff
@@ -0,0 +1,710 @@
+    type Parameters = T;
+    type MerklePath = FieldBasedMHTPath<T>;
+
+    fn append(&mut self, leaf: T::Data) -> &mut Self {
+        if self.new_elem_pos[0] < self.final_pos[0] {
+            self.array_nodes[self.new_elem_pos[0]] = leaf;
+            self.new_elem_pos[0] += 1;
+        }
+
+        if self.new_elem_pos[0] == self.final_pos[0] {
+            self.compute_subtree();
+        }
+
+        if (self.new_elem_pos[0] - self.processed_pos[0]) >= self.processing_step {
+            self.compute_subtree();
+        }
+        self
+    }
+
+    fn finalize(&self) -> Self {
+        let mut copy = (*self).clone();
+        copy.new_elem_pos[0] = copy.final_pos[0];
+        copy.compute_subtree();
+        copy.finalized = true;
+        copy.root = *copy.array_nodes.last().unwrap();
+        copy
+    }
+
+    fn finalize_in_place(&mut self) -> &mut Self {
+        self.new_elem_pos[0] = self.final_pos[0];
+        self.compute_subtree();
+        self.finalized = true;
+        self.root = *self.array_nodes.last().unwrap();
+        self
+    }
+
+    fn reset(&mut self) -> &mut Self {
+        for i in 0..self.new_elem_pos.len() {
+            self.new_elem_pos[i] = self.initial_pos[i];
+            self.processed_pos[i] = self.initial_pos[i];
+        }
+        self.finalized = false;
+
+        self
+    }
+
+    fn root(&self) -> Option<T::Data> {
+        match self.finalized {
+            true => Some(self.root.clone()),
+            false => None
+        }
+    }
+
+    fn get_merkle_path(&self, leaf_index: usize) -> Option<Self::MerklePath> {
+        match self.finalized {
+            true => {
+                let num_leaves = T::MERKLE_ARITY.pow(self.height as u32);
+                let mut merkle_path = Vec::with_capacity(self.height);
+
+                let mut node_index = leaf_index;
+                for _ in 0..self.height {
+                    let mut siblings = Vec::with_capacity(T::MERKLE_ARITY - 1);
+
+                    // Based on the index of the node, we must compute the index of the left-most children
+                    let start_position = node_index - ( node_index % T::MERKLE_ARITY );
+
+                    // Then, the right most children index is simply given by adding the arity
+                    let end_position = start_position + T::MERKLE_ARITY;
+
+                    // We must save the siblings of the actual node
+                    for i in start_position..end_position {
+                        if i != node_index {
+                            siblings.push(self.array_nodes[i])
+                        }
+                    }
+
+                    // Remember to normalize the node_index with respect to the length of siblings
+                    merkle_path.push((siblings, node_index % T::MERKLE_ARITY));
+
+                    // Get parent index for next iteration
+                    node_index = num_leaves + (node_index/T::MERKLE_ARITY);
```

Comment:
after that, shouldn't we update num_leaves according to the next level?
num_leaves = num_leaves/MERKLE_ARITY?

Reply:
Ah, no. I think it is right, but a bit "encrypted" as it is not easy to understand the role of num_leaves, which is constant.

Why not simply doing 
```
node_index = initial_pos[next level] + (node_index-initial_pos[current level])/MERKLE_ARITY
```